### PR TITLE
Add Question OOI form rendering on the object detail page

### LINKED ISSUE #288: Question OOIs with json-schema
See #130 
Draft PR: #1394

Some considerations:

- We want questions to be formulated by json-schemas and answers as json
- In the frontend we can create forms from json-schema's and json from form submissions
- ~We might need to generate id's for questions~ Unique Questions per OOI are probably fine?
- Schema's are defined in Bits
- Per answer, the right Config needs to be created. Determining which Config can be done through:
  - **The mime types, by adding an id per schema (best practice for json-schema's) and triggering the right normalizer**. However, this would also be communicating the bit_id for Configs, so we are somewhat abusing notation here.
  - ~Adding a identifier field in the answer~
  - ~Adding an argument (but I'm not a fan of the mostly unused arguments field, see https://github.com/minvws/nl-kat-coordination/issues/763)~
  - ~Try to infer it through the input_ooi~ This will not work if we start attaching multiple configs to a Network.

Still, Rocky (saving the raw file) needs to know what the "type" of Question is, so somehow the Question object (either as a field or in its schema) needs to carry an identifier.

Questions:
- Do we need to add a new page with an overview? Quickly seeing unanswered question would be nice I'd say. 

Config flow and cardinality:

```mermaid
classDiagram

direction RL
    class Question
    class Config
    class JsonAnswer
    class JsonAnswer2["JsonAnswer"]

    Question : ooi[OOI]
    Question : schema[str]

    Config : ooi[OOI]
    Config : config[str]
    Config : bit_id[str]

    OOI <|-- Question
    Question <|-- JsonAnswer
    Question <|-- JsonAnswer2
    JsonAnswer <|-- Config
    JsonAnswer2 <|-- Config
```

Resulting ERD:

```mermaid
classDiagram

direction RL
    class Question
    class Config
    class OOI

    Question : ooi[OOI]
    Question : schema[str]

    Config : ooi[OOI]
    Config : config[str]
    Config : bit_id[str]

    OOI <|-- Question
    OOI <|-- Config
```

#### The Flow

```mermaid
sequenceDiagram
    participant User
    participant Rocky
    participant Normalizer
    participant Octopoes
    participant Bits
    participant Bytes

    Normalizer->>Octopoes: Add Network
    Bits->>Octopoes: Add Question["What ports may be open for this Network?"]
    Rocky->>Octopoes: Get Question
    Rocky->>User: Prompt Question to user
    User->>Rocky: Give answer (form)
    Rocky->>Bytes: Add answer (json) to Bytes
    Bytes->>Normalizer: Read answer
    Normalizer->>Octopoes: Create Config
    Bits->>Octopoes: Read Config
```

---

### Generating Question (forms)

To create question OOI's:
- Rebuild Rocky and the boefjes/normalizers (Or: make kat)
- Add a Network object
- Increase its clearance level to at least L2
- Wait a few seconds and refresh the objects page (perhaps filter on Question objects)
- Open the Question detail page and see the json form rendered.

### PR DESCRIPTION
**Pull request by Donnype**
### Changes
This adds the rendering of questions and saving answers to bytes in rocky. Note: this merges into https://github.com/minvws/nl-kat-coordination/pull/1407 for now (PR split up) and depends on it.

### Issue link

(Partially) closes #288

### Proof
image

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [x] I have written unit tests for the changes or fixes I made.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [ ] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/rocky/views/ooi_detail.py [hunk 1]
```diff
@@ -70,10 +64,38 @@ def handle_page_action(self, action: str) -> bool:
...
                 ooi = self.get_single_ooi(pk=ooi_id)
                 self.run_boefje_for_oois(boefje, [ooi])
-                return True
 
+                success_message = (
+                    "Your scan is running successfully in the background. \n "
+                    "Results will be added to the object list when they are in. "
+                    "It may take some time, a refresh of the page may be needed to show the results."
+                )
+                messages.add_message(self.request, messages.SUCCESS, success_message)
```
Comment (ammar92) [inline rocky/rocky/views/ooi_detail.py:73]: This message is not translated

File: rocky/rocky/views/ooi_detail.py [hunk 2]
```diff
@@ -70,10 +64,38 @@ def handle_page_action(self, action: str) -> bool:
...
+                return redirect("task_list", organization_code=self.organization.code)
+
+            if action == PageActions.SUBMIT_ANSWER.value:
+                if not isinstance(self.ooi, Question):
+                    messages.add_message(self.request, messages.ERROR, "Only Question OOIs can be answered.")
```
Comment (ammar92) [inline rocky/rocky/views/ooi_detail.py:78]: Not translated

File: rocky/tests/conftest.py [hunk 3]
```diff
@@ -450,8 +451,12 @@ def mock_scheduler(mocker):
     return mocker.patch("rocky.views.ooi_detail.scheduler.client")
 
 
-def get_boefjes_data():
-    return json.loads((Path(__file__).parent / "stubs" / "katalogus_boefjes.json").read_text())
+def get_stub_path(file_name: str) -> Path:
+    return Path(__file__).parent / "stubs" / file_name
```
Comment (ammar92) [inline rocky/tests/conftest.py:455]: Might as well be a static variable or inlined in the function below

File: rocky/rocky/views/ooi_detail.py [hunk 4]
```diff
@@ -70,10 +64,38 @@ def handle_page_action(self, action: str) -> bool:
...
+
+                if not validator.is_valid(parsed_schema_answer):
+                    for error in validator.iter_errors(parsed_schema_answer):
+                        messages.add_message(self.request, messages.ERROR, error.message)
+
+                    return self.get(self.request, status_code=500, *self.args, **self.kwargs)
+
+                self.bytes_client.upload_raw(schema_answer, {"answer", f"{self.ooi.schema_id}"}, self.ooi.reference)
+                messages.add_message(self.request, messages.SUCCESS, "Question has been answered.")
+                return self.get(self.request, status_code=201, *self.args, **self.kwargs)
```
Comment (ammar92) [inline rocky/rocky/views/ooi_detail.py:93]: I'm not sure if the `status_code` argument does anything. I haven't seen this in Rocky yet and Django doesn't seem to know it. Did you actually get one of these (non-200) status codes rendered in this view?

File: rocky/tests/objects/test_objects_detail.py [hunk 5]
```diff
@@ -63,6 +84,33 @@ def test_ooi_detail(
     assertContains(response, "Completed")
     assertContains(response, "Created date")
     assertContains(response, "9, 2022, 11:53 a.m.")
+    assertNotContains(response, "Question")
+    assertNotContains(response, "Rendered Question Form")
+
+
+def test_question_detail(
```
Comment (ammar92) [inline rocky/tests/objects/test_objects_detail.py:91]: You might want to test one of the error paths (500s) or the created response (201)

**Review by ammar92 — COMMENTED** [review]
Looks good in general 👍 Just a few remarks:
- check the `status_code` argument
- perhaps one of the front-end team members can check the js asset for correctness and style?
- at least one tests that tests the unhappy path(s)

File: rocky/tests/conftest.py (hunk 3 above)
Comment (Donnype) [reply rocky/tests/conftest.py:456]: I'm using it elsewhere as well though

File: rocky/rocky/views/ooi_detail.py (hunk 4 above)
Comment (Donnype) [reply rocky/rocky/views/ooi_detail.py:93]: This functionality was based on previous investigation and work (and feedback) at the time about that we want to return 400 error on bad input as well. It actually works as well! Here's e.g. the 201:

Schermafbeelding 2023-07-25 om 15 54 10

File: rocky/tests/objects/test_objects_detail.py (hunk 5 above)
Comment (Donnype) [reply rocky/tests/objects/test_objects_detail.py:91]: Done!

**Comment by Donnype** [comment]
I created [a separate ticket](https://github.com/minvws/nl-kat-coordination/issues/1457) for this week to write some documentation.

**Comment by Donnype** [comment]
> Looks good in general 👍 Just a few remarks:
> 
> * perhaps one of the front-end team members can check the js asset for correctness and style?

@HeleenSG @TwistMeister ?

**Comment by Darwinkel** [comment]
## Checklist for QA:
- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
    - [x] I have managed to pass the onboarding flow
    - [x] Objects and Findings are created properly
    - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.

### What works:
* Works much better than entering a json blob :)
* No apparent regressions

### What doesn't work:
* No new ConfigOOI is created (yet? I'm assuming that's a follow-up PR)

### Bug or feature?:
* Not sure if showing the JSON schema on the object details page is smart. It's almost certainly unintelligible for the average KAT user
image

**Comment by Donnype** [comment]
@Darwinkel Correct about the ConfigOOI, and agreed on the showing of the JSON schema on the object details page. Not sure about a good resolution that does not open a pool of questions about specific-ooi-field-rendering..

**Comment by underdarknl** [comment]
Agreed about the json being vissible. It should be an easy fix in je JS to hide those lines.

**Review by underdarknl — APPROVED** [review]

**Comment by Donnype** [comment]
@Darwinkel @underdarknl Fixed

**Pull request merged**